### ISSUE TITLE
[BUG]: PVC provisioning on SRDF Metro Storage Class suddenly stopped working

### ISSUE THREAD

**Issue by philrich** [dell/csm#748]
### Bug Description

Suddenly it is impossible to provision new PVCs as Metro/RDF Devices. We use this StorageClass:
```
kind: StorageClass
apiVersion: storage.k8s.io/v1
metadata:
  name: powermax-x
provisioner: csi-powermax.dellemc.com
parameters:
  csi.storage.k8s.io/fstype: xfs
  SYMID: <symid-1>
  replication.storage.dell.com/RemoteSRP: SRP_1
  replication.storage.dell.com/isReplicationEnabled: 'true'
  ServiceLevel: Diamond
  SRP: SRP_1
  replication.storage.dell.com/RdfGroup: '150'
  replication.storage.dell.com/RdfMode: METRO
  replication.storage.dell.com/RemoteSYMID: <symid-2>
  replication.storage.dell.com/RemoteRDFGroup: '150'
  replication.storage.dell.com/RemoteServiceLevel: Diamond
  replication.storage.dell.com/Bias: 'true'
reclaimPolicy: Delete
allowVolumeExpansion: true
volumeBindingMode: Immediate
```
Following Error appears in the controller log:
```
{"level":"info","msg":"/csi.v1.Controller/CreateVolume: REP 0824: rpc error: code = Internal desc = Could not create storage group replica for (csi-rep-sg-150-METRO) - Error (A problem occurred creating the Storage Group SRDF Group resource: The operation is not allowed for a device in an active RDF/Metro session)","time":"2023-04-06T15:02:51.238954202Z"}
```
This happens in both of our clusters (Other Cluster uses 140 as SRDF Group).
Only way to fix this (in our Test Cluster) was to remove the SRDF Group (and all Volumes it contained) and recreate it. After that PVC creation worked again.

##### OpenShift Powermax Controller Log

```
{"level":"info","msg":"/csi.v1.Controller/CreateVolume: REQ 0824: Name=pmax-a97c233c4f, CapacityRange=required_bytes:8589934592 , VolumeCapabilities=[mount:\u003cfs_type:\"xfs\" \u003e access_mode:\u003cmode:SINGLE_NODE_MULTI_WRITER \u003e ], Parameters=map[SRP:SRP_1 SYMID:<symid-1> ServiceLevel:Diamond csi.storage.k8s.io/pv/name:pmax-a97c233c4f csi.storage.k8s.io/pvc/name:data-keycloak-postgresql-0 csi.storage.k8s.io/pvc/namespace:cms-et-app replication.storage.dell.com/Bias:true replication.storage.dell.com/RdfGroup:150 replication.storage.dell.com/RdfMode:METRO replication.storage.dell.com/RemoteRDFGroup:150 replication.storage.dell.com/RemoteSRP:SRP_1 replication.storage.dell.com/RemoteSYMID:<symid-2> replication.storage.dell.com/RemoteServiceLevel:Diamond replication.storage.dell.com/isReplicationEnabled:true], AccessibilityRequirements=requisite:\u003csegments:\u003ckey:\"csi-powermax.dellemc.com/<symid-1>\" value:\"csi-powermax.dellemc.com\" \u003e segments:\u003ckey:\"csi-powermax.dellemc....
{"level":"info","msg":"Failing over to the array: <symid-2>","time":"2023-04-06T15:02:46.777574577Z"}
{"level":"info","msg":"StoragePoolCapacities: \u0026v100.SrpCap{SubAllocCapInTB:61.93, SubTotInTB:131.03, SnapModInTB:0.06, SnapTotInTB:2.05, UsableUsedInTB:45.55, UsableTotInTB:95.78, EffectiveUsedCapacityPercent:46}","time":"2023-04-06T15:02:47.942450624Z"}
{"level":"info","msg":"totalSrcCapInGB 98078.720000 usedSrpCapInGB 46643.200000 remainingCapInGB 51435.520000 maxAvailBytes 55227910717440","time":"2023-04-06T15:02:47.942519063Z"}
{"level":"info","msg":"StoragePoolCapacities: \u0026v100.SrpCap{SubAllocCapInTB:66.82, SubTotInTB:136.24, SnapModInTB:0.06, SnapTotInTB:2.05, UsableUsedInTB:48.13, UsableTotInTB:95.77, EffectiveUsedCapacityPercent:49}","time":"2023-04-06T15:02:47.955536258Z"}
{"level":"info","msg":"totalSrcCapInGB 98068.480000 usedSrpCapInGB 49285.120000 remainingCapInGB 48783.360000 maxAvailBytes 52380347400192","time":"2023-04-06T15:02:47.955562758Z"}
{"Accessibility":{"requisite":[{"segments":{"csi-powermax.dellemc.com/<symid-1>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-1>.fc":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>.fc":"csi-powermax.dellemc.com"}}],"preferred":[{"segments":{"csi-powermax.dellemc.com/<symid-1>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-1>.fc":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>.fc":"csi-powermax.dellemc.com"}}]},"ApplicationPrefix":"","CSIRequestID":"824","LocalProtectionGroupID":"csi-rep-sg-150-METRO","LocalRDFGroup":"150","RemoteProtectionGroupID":"csi-rep-sg-150-METRO","RemoteRDFGroup":"150","RemoteSymmID":"<symid-2>","ReplicationEnabled":"true","SRDFMode":"METRO","SRP":"SRP_1","SymmetrixID":"<symid-1>","level":"info","msg":"Executing CreateVolume with following fields","requiredCylinders":43...
{"level":"debug","msg":"Requesting a lock for csi-rep-sg-150-METRO<symid-1> with requestID: 824 at: 2023-04-06 15:02:47.955619965 +0000 UTC m=+182230.008984404","time":"2023-04-06T15:02:47.955625335Z"}
{"level":"debug","msg":"Acquired - Lock Number:3443866, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1> at: 2023-04-06 15:02:47.95563833 +0000 UTC m=+182230.009002769","time":"2023-04-06T15:02:47.955641015Z"}
{"level":"debug","msg":"Released - Lock Number: 3443866, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1>","time":"2023-04-06T15:02:48.068646053Z"}
{"level":"debug","msg":"Calling GetVolumeIDList for idempotency test","time":"2023-04-06T15:02:48.179710359Z"}
{"Accessibility":{"requisite":[{"segments":{"csi-powermax.dellemc.com/<symid-1>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-1>.fc":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>.fc":"csi-powermax.dellemc.com"}}],"preferred":[{"segments":{"csi-powermax.dellemc.com/<symid-1>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-1>.fc":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>":"csi-powermax.dellemc.com","csi-powermax.dellemc.com/<symid-2>.fc":"csi-powermax.dellemc.com"}}]},"ApplicationPrefix":"","CSIRequestID":"824","LocalProtectionGroupID":"csi-rep-sg-150-METRO","LocalRDFGroup":"150","RemoteProtectionGroupID":"csi-rep-sg-150-METRO","RemoteRDFGroup":"150","RemoteSymmID":"<symid-2>","ReplicationEnabled":"true","SRDFMode":"METRO","SRP":"SRP_1","SymmetrixID":"<symid-1>","level":"info","msg":"Calling GetVolumeByID for idempotence check","requiredCylinders":437...
{"level":"error","msg":"Failed to fetch rdf pair information for (00601) - Error (The device must be an RDF device for this function)","time":"2023-04-06T15:02:48.369410149Z"}
{"level":"debug","msg":"Local Volume already exist, skipping creation (00601)","time":"2023-04-06T15:02:48.369467597Z"}
{"level":"debug","msg":"Creating meta data for HTTP header","time":"2023-04-06T15:02:48.369476233Z"}
{"level":"debug","msg":"RDF: Found Rdf enabled","time":"2023-04-06T15:02:48.369496822Z"}
{"level":"debug","msg":"Requesting a lock for csi-rep-sg-150-METRO<symid-1> with requestID: 824 at: 2023-04-06 15:02:48.369504316 +0000 UTC m=+182230.422868755","time":"2023-04-06T15:02:48.369514806Z"}
{"level":"debug","msg":"Acquired - Lock Number:7352393, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1> at: 2023-04-06 15:02:48.369539673 +0000 UTC m=+182230.422904112","time":"2023-04-06T15:02:48.369544612Z"}
{"level":"info","msg":"payload: {\"editStorageGroupActionParam\":{\"expandStorageGroupParam\":{\"addSpecificVolumeParam\":{\"volumeId\":[\"00601\"],\"remoteSymmSGInfoParam\":{\"remote_symmetrix_1_id\":\"<symid-2>\",\"remote_symmetrix_1_sgs\":[\"csi-rep-sg-150-METRO\"]}}}},\"executionOption\":\"SYNCHRONOUS\"}","time":"2023-04-06T15:02:48.369578265Z"}
{"level":"debug","msg":"volume (00601) added to protected SG (csi-rep-sg-150-METRO)","time":"2023-04-06T15:02:49.354686522Z"}
{"level":"debug","msg":"Released - Lock Number: 7352393, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1>","time":"2023-04-06T15:02:49.354765551Z"}
{"level":"debug","msg":"Requesting a lock for csi-rep-sg-150-METRO<symid-1> with requestID: 824 at: 2023-04-06 15:02:49.354777514 +0000 UTC m=+182231.408141943","time":"2023-04-06T15:02:49.354790167Z"}
{"level":"debug","msg":"Acquired - Lock Number:9903394, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1> at: 2023-04-06 15:02:49.354811187 +0000 UTC m=+182231.408175616","time":"2023-04-06T15:02:49.354815154Z"}
{"level":"debug","msg":"RDF: rdfg has 8 devices ! for vol(00601)","time":"2023-04-06T15:02:49.645068852Z"}
{"level":"debug","msg":"Can not found Remote storage group, proceed","time":"2023-04-06T15:02:49.655987338Z"}
{"level":"info","msg":"payload: {\"remoteSymmId\":\"<symid-2>\",\"replicationMode\":\"Active\",\"rdfgNumber\":150,\"forceNewRdfGroup\":\"\",\"establish\":true,\"metroBias\":true,\"remoteStorageGroupName\":\"csi-rep-sg-150-METRO\",\"thinPool\":\"\",\"fastPolicy\":\"\",\"remoteSLO\":\"\",\"noCompression\":false,\"executionOption\":\"SYNCHRONOUS\"}","time":"2023-04-06T15:02:49.656016854Z"}
{"level":"error","msg":"Could not create storage group replica for (csi-rep-sg-150-METRO)","time":"2023-04-06T15:02:50.706276099Z"}
{"level":"debug","msg":"Released - Lock Number: 9903394, requestID: 824, resourceID: csi-rep-sg-150-METRO<symid-1>","time":"2023-04-06T15:02:50.706327286Z"}
{"level":"error","msg":"Proceeding to remove volume from protected storage group as rollback","time":"2023-04-06T15:02:50.706337064Z"}
{"level":"info","msg":"payload: {\"editStorageGroupActionParam\":{\"removeVolumeParam\":{\"volumeId\":[\"00601\"],\"remoteSymmSGInfoParam\":{\"force\":true}}},\"executionOption\":\"SYNCHRONOUS\"}","time":"2023-04-06T15:02:50.706368263Z"}
{"level":"info","msg":"Successfully removed volumes: [00601] from SG: csi-rep-sg-150-METRO","time":"2023-04-06T15:02:51.238856327Z"}
{"level":"info","msg":"/csi.v1.Controller/CreateVolume: REP 0824: rpc error: code = Internal desc = Could not create storage group replica for (csi-rep-sg-150-METRO) - Error (A problem occurred creating the Storage Group SRDF Group resource: The operation is not allowed for a device in an active RDF/Metro session)","time":"2023-04-06T15:02:51.238954202Z"}
```
##### Unisphere - /usr/emc/API/symapi/log/symapi-20230406.log
```
04/06/2023 16:07:05.944   2809    217 EMC:UNIVMAX                emcStorGrpDevAdd     STARTING on csi-rep-sg-150-METRO(Storage group) 'ADD' operation. Symm=<symid-1> Flags=0x0
04/06/2023 16:07:05.948   2809    217 EMC:UNIVMAX                storgrp_audit_start  Devices: [ 0601 ]
04/06/2023 16:07:05.952   2809    217 EMC:UNIVMAX                grp_audit_exit       On Storage group csi-rep-sg-150-METRO operation 'ADD' SUCCESSFULLY COMPLETED
04/06/2023 16:07:07.918   2809    217 EMC:UNIVMAX                evaluate_dev_rdf_sta Cannot add devices to RDF group in active SRDF/Metro session unless format or exempt option is specified
04/06/2023 16:07:07.918   2809    217 EMC:UNIVMAX                SymDevListRdfControl grp: NA  ldev:  actionP: CREATEPAIR - The operation is not allowed for a device in an active RDF/Metro session (local sid/dskno <symid-1>/00601) (remote sid/dskno <symid-2>/00603)
04/06/2023 16:07:08.269   2809    217 EMC:UNIVMAX                emcStorGrpDevRemove  STARTING on csi-rep-sg-150-METRO(Storage group) 'REMOVE' operation. Symm=<symid-1> Flags=0x0
04/06/2023 16:07:08.273   2809    217 EMC:UNIVMAX                storgrp_audit_start  Devices: [ 0601 ]
04/06/2023 16:07:08.278   2809    217 EMC:UNIVMAX                grp_audit_exit       On Storage group csi-rep-sg-150-METRO operation 'REMOVE' SUCCESSFULLY COMPLETED
```
(don't mind the timestamps - it loops forever.)

### Additional Environment Information

We have 2 Unisphere Installations behind the CSI Reverse Proxy

### Steps to Reproduce

It happened after some time.
Unknown how to reproduce.

### Expected Behavior

PVC creation should be sucessfull

### CSM Driver(s)

CSI Driver for Powermax v2.5.0

### Installation Type

Dell CSI Operator v1.10.0 (from Redhat Certified Community)

### Container Orchestrator

OpenShift 4.10.50

### Operating System

RHCOS

**Comment by csmbot** [comment]
@philrich: Thank you for submitting this issue!

The issue is currently awaiting triage. Please make sure you have given us as much context as possible.

If the maintainers determine this is a relevant issue, they will remove the needs-triage label and assign an appropriate priority label.

------------------------------------------------------------------------------------------------------------------------
We want your feedback! If you have any questions or suggestions regarding our contributing process/workflow, please reach out to us at container.storage.modules@dell.com.

**Comment by delldubey** [comment]
Hi,
To investigate further, please share the following details:

- What is the scale of your application, like how many volumes?
- For how many volumes are you observing this? 
- Can you update the driver with **X_CSI_POWERMAX_DEBUG**, sets this to true, and then share the new logs?

Generally, these issue happens when the load on Unisphere is more than it can handle.

Thanks.

**Comment by philrich** [comment]
Hi,
Here are your answers:
- We are currently in the process of migrating the applications from ovirt-csi to dell csi driver. At this time only one application (7 volumes per instance) needs dell storage. But - by the end of the year we want to go all in and migrate all of our RHV VMs to Openshift Virtualization - this would be > 500 volumes.
We just started evaluating dell csi driver two months ago an all went well.
- As soon as the above Error is seen we are not able to provision any additional PVC on our RDF/Metro storage class. All in all we have 3 storage classes: powermax-a (no rdf, only on storage in datacenter a), powermax-b (same, datacenter b) and powermax-x. We still can provision PVCs on powermax-a/b even after powermax-x failed. But for availability reasons we want most of our PVCs to be mirrored.
- Did this, see attached log file: 
[powermax-controller-d8466749-p8s6q-driver.log](https://github.com/dell/csm/files/11183030/powermax-controller-d8466749-p8s6q-driver.log)

Regarding load on Unisphere: We have two installations (one in each datacenter), we updated them to the latest 10.x.x.x version after the problem occured but that did not help. At least the unix load is < 0.1. In symapi*.log, only openshift driver logs from our 2 clusters are seen, so i would assume its mostly idle. Unisphere ist installed on baremetal 6 Core Intel CPU, 32 GB RAM

FYI: we also planned to upgrade the dell-csi-operator to the recently released v1.11 (powermax v2.6) but did not yet succeed as our OpenShift installation is air gapped and we had problems to mirror the content from docker.io.

**Comment by delldubey** [comment]
Hi
Thanks for the logs, we will go through this,
Meanwhile, you can leverage the Auto-creation of RDF Group [usage](https://dell.github.io/csm-docs/docs/replication/high-availability/#usage), to bypass the issue till we come back. 
You can [update ](https://github.com/dell/csi-powermax/blob/main/samples/storageclass/powermax_srdf.yaml#L78)your storage class to let CSI Driver create its own RDF Group for its use. 
This new feature may help you with your issue.

Thanks.

**Comment by philrich** [comment]
Hi,
The SRDF Group 150 was created exclusively for our OpenShift Staging Cluster. All volumes in this group were created by the csi driver itself. Everything worked as expected until last Monday, so something broke and is preventing provisioning of future volumes.

Why would Auto-SRDF be different in this case? As far as I understand, just the group number would be different, counting down from 250.

I reexamined the controller log from the time the error first occured and I see another error shortly before first appearance of the SRDF problem:
```
{"PUT":"univmax/restapi/100/sloprovisioning/symmetrix/<symid-1>/volume/0060B-0060D","VolumeID":"0060B-0060D","level":"error","msg":"Error in DeleteVolume: A problem occurred deleting the volume resources: The requested operation cannot be performed on an enabled CG member","time":"2023-04-03T14:08:55.79993126Z"
```
I have attached the full log here: [powermax-controller-c4bf9b79d-j7zzg-driver.log](https://github.com/dell/csm/files/11200383/powermax-controller-c4bf9b79d-j7zzg-driver.log)
I remember, we had some stuck _DEL volumes which we deleted manually after the driver gave up, maybe this is related?

- First occurence of delete error: Line 3750
- First occurence of srdf problem: Line 3845
